File: src/CoverageBin.py
#!/usr/bin/env python

"""
Created: September 21,2017

This file takes input a HiC-pro pipeline output of validpairs file, Chromosome length file, and a peak detection file
it generates binned intervals from the input chromosomes
and creates coverage vector for individual chromosomes
Dependin on the input peak file, it also marks individual binned segments as either peak or non-peak

Author: Sourya Bhattacharyya
Vijay-Ay lab, LJI
"""

import sys
import os
from optparse import OptionParser

# used to open the gzip compressed text file
import gzip 

#==========================================================
""" 
this class defines a particular chromosome interval
indexed by (chromosome, start, end) - dictionary key 
"""
class ChrmSegment(object):  
    def __init__(self):
        self.is_peak = False
        self.read_count = 0

    def _CheckIfPeak(self):
        return self.is_peak

    def _GetReadCount(self):
        return self.read_count

    def _IncrementRead(self, count=1):
        self.read_count = self.read_count + count

    def _SetPeak(self):
        self.is_peak = True

#===============================================
def main():
    parser = OptionParser() #(usage=usage)
    parser.add_option("-p", "--peakfile", dest="peakfile", help="Peak detection file")
    parser.add_option("-i", "--inpfile", dest="inpfile", help="Input valid pairs file (from HiC-pro pipeline)")
    parser.add_option("-b", "--binsize", dest="binsize", type="int", help="Size of bins employed. DEFAULT 5000 (indicating 5 Kb).")
    parser.add_option("-o", "--outfile", dest="outfile", help="output file storing the coverage of individual genomic bins")
    parser.add_option("-c", "--chrsizefile", dest="chrsizefile", help="File containing chromosome size information")
    parser.set_defaults(peakfile=None, inpfile=None, binsize=5000, outfile=None, chrsizefile=None)
    (options, args) = parser.parse_args()

    # global variables from command line options
    global inpfile
    global peakfile
    global outfile
    global bin_size
    global chrsizefile

    # dictionary structure which stores the instances of the class ChrmSegment
    global SegmentDict
    SegmentDict = dict()

    # two lists which store the chromosome name and their length information
    # with respect to the input chromosome length file
    ChrNameList_Ref = []
    ChrLenList_Ref = []

    # list storing the chromosome names provided in the input validpairs file
    ChrNameList_Current = [] 

    if options.inpfile is not None:
        inpfile = options.inpfile
    else:
        sys.exit("Input validpairs file (output of HiC-pro pipeline) is not provided - quit !!")

    if options.peakfile is not None:
        peakfile = options.peakfile
    else:
        sys.exit("Peak detection file is not provided - quit !!")

    if options.outfile is not None:
        outfile = options.outfile
    else:
        sys.exit("Output file for storing the coverage of segments is not provided - quit !!")

    if options.chrsizefile is not None:
        chrsizefile = options.chrsizefile
    else:
        sys.exit("Chromosome size file is not provided - quit !!")

    # input bin size parameter  
    bin_size = int(options.binsize)

    #=====================================================
    # first read the reference chromosome length file
    # and assign the name and lengths of individual chromosomes
    #=====================================================
    with open(chrsizefile, 'r') as fp:
        for line in fp:
            linecontents = (line.rstrip()).split()
            curr_chr = linecontents[0]
            curr_chr_size = int(linecontents[1])
            ChrNameList_Ref.append(curr_chr)
            ChrLenList_Ref.append(curr_chr_size)

    #=====================================================
    # now read the input valid pairs file
    # and compute the coverage information
    # with respect to 'cis' reads
    #=====================================================
    with gzip.open(inpfile,'r') as fin:    
        for line in fin:
            contents = line.split()
            # check cis interaction
            if (str(contents[1]) == str(contents[4])):
                chrname = str(contents[1])
                read1_pos = int(contents[2])
                read2_pos = int(contents[5])
                
                # check if the chromosome is not in the 
                # already processed list
                # in such a case, include this chromosome
                # and also create the binned intervals
                if chrname not in ChrNameList_Current:
                    # this chromosome is not yet processed
                    ChrNameList_Current.append(chrname)
                    # create python dictionary 
                    # and the bin size specific intervals for this chromosome
                    chrsize = ChrLenList_Ref[ChrNameList_Ref.index(chrname)]
                    if ((chrsize % bin_size) == 0):
                        interval_end = chrsize
                    else:
                        interval_end = (int(chrsize / bin_size) + 1) * bin_size
                    for val in range(0, interval_end, bin_size):
                        curr_key = (chrname, val, (val + bin_size))
                        SegmentDict.setdefault(curr_key, ChrmSegment())
                
                # now process the current paired end read
                # read position 1
                bin_start1 = (int(read1_pos / bin_size)) * bin_size
                # increment the read count in the corresponding dictionary entry of this bin
                curr_key1 = (chrname, bin_start1, (bin_start1 + bin_size))
                SegmentDict[curr_key1]._IncrementRead()
                # read position 2
                bin_start2 = (int(read2_pos / bin_size)) * bin_size
                # increment the read count in the corresponding dictionary entry of this bin
                curr_key2 = (chrname, bin_start2, (bin_start2 + bin_size))
                SegmentDict[curr_key2]._IncrementRead()
    
    #=====================================================
    # scan the peak input file (if provided) and mark the corresponding 
    # chromosome intervals as 1 (having peak)
    #=====================================================
    with open(peakfile, 'r') as fp:
        for line in fp:
            linecontents = (line.rstrip()).split()
            curr_chr = linecontents[0]
            peak_start = int(linecontents[1])
            peak_end = int(linecontents[2])
            # only process those peaks whose chromosome 
            # is present in the current valid pairs file
            # this is possible when the peak information is downloaded from a reference
            if curr_chr in ChrNameList_Current:
                interval_start = (int(peak_start / bin_size)) * bin_size
                if ((peak_end % bin_size) == 0):
                    interval_end = peak_end
                else:
                    interval_end = (int(peak_end / bin_size) + 1) * bin_size
                for val in range(interval_start, interval_end, bin_size):
                    curr_key = (curr_chr, val, (val + bin_size))
                    # mark the corresponding dictionary entry as a peak segment
                    # since it has an overlap with the MACS2 derived peak intervals
                    SegmentDict[curr_key]._SetPeak()
    
    # now write the genomic intervals and corresponding read count + peak information
    fp_cov = open(outfile, 'w')
    fp_cov.write('Chr' + '\t' + 'Start' + '\t' + 'End' + '\t' + 'Coverage' + '\t' + 'IsPeak')
    for i in range(len(ChrNameList_Current)):
        curr_chr = ChrNameList_Current[i]
        curr_chr_size = ChrLenList_Ref[ChrNameList_Ref.index(curr_chr)]
        if ((curr_chr_size % bin_size) == 0):
            interval_end = curr_chr_size
        else:
            interval_end = (int(curr_chr_size / bin_size) + 1) * bin_size
        for val in range(0, interval_end, bin_size):
            curr_key = (curr_chr, val, (val + bin_size))
            fp_cov.write('\n' + str(curr_chr) + '\t' + str(val) + '\t' + str(val + bin_size) + '\t' + str(SegmentDict[curr_key]._GetReadCount()) + '\t' + str(int(SegmentDict[curr_key]._CheckIfPeak())))
    fp_cov.close()

    return

#===============================================
if __name__ == "__main__":
    main()

File: src/CoverageBin.r
#!/usr/bin/env Rscript

library(optparse)
library(GenomicRanges)
library(tools)
library(data.table)

options(scipen = 999)
options(datatable.fread.datatable=FALSE)

#======================
# script to compute 1D bin specific HiChIP coverage
# and also their overlap with reference ChIP-seq / HiChIP peak files
#======================

Overlap1D <- function(Inpdata1, Inpdata2, boundary=1, offset=0, uniqov=TRUE) {

	ov1 <- as.data.frame(findOverlaps(GRanges(Inpdata1[,1], IRanges(Inpdata1[,2]+boundary-offset, Inpdata1[,3]-boundary+offset)),GRanges(Inpdata2[,1], IRanges(Inpdata2[,2]+boundary-offset, Inpdata2[,3]-boundary+offset))))
	if (uniqov == TRUE) {
		ov_idx_file1 <- unique(ov1[,1])
		ov_idx_file2 <- unique(ov1[,2])		
	} else {
		ov_idx_file1 <- ov1[,1]
		ov_idx_file2 <- ov1[,2]
	}
	nonov_idx_file1 <- setdiff(seq(1, nrow(Inpdata1)), ov_idx_file1)
	nonov_idx_file2 <- setdiff(seq(1, nrow(Inpdata2)), ov_idx_file2)

	# return the overlapping and non-overlapping set of indices
	newList <- list(A_AND_B = ov_idx_file1, B_AND_A = ov_idx_file2, A_MINUS_B = nonov_idx_file1, B_MINUS_A = nonov_idx_file2)
	return(newList)

}

#====================================================
option_list = list(
  	make_option(c("--InpFile"), type="character", default=NULL, help="Input file with locus pairs and their contact count."),
  	make_option(c("--PeakFile"), type="character", default=NULL, help="Reference ChIP-seq or HiChIP peak file."),
  	make_option(c("--BinSize"), type="integer", action="store", default=5000, help="Bin size employed. Default 5000 (5 Kb)."),
  	make_option(c("--ChrSizeFile"), type="character", default=NULL, help="File containing chromosome sizes corresponding to the reference genome."),
  	make_option(c("--OutFile"), type="character", default=NULL, help="Output file storing 1D bin specific HiChIP coverage along with their overlap with input peak segments.")
); 
 
opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);

OutDir <- dirname(opt$OutFile)

# counter of valid chromosomes (containing one or more interactions)
valid_chr_count <- 0

# first use the reference chromosome size file and input bin size
# to extract all distinct 1D bins
temp_1D_binfile <- paste0(OutDir, '/temp_1D_ALLBins.bed')
system(paste("bedtools makewindows -g", opt$ChrSizeFile, "-w", opt$BinSize, ">", temp_1D_binfile))

# read the chromosomes of input ChrSizeFile
ChrData <- read.table(opt$ChrSizeFile, header=F, sep="\t", stringsAsFactors=F)
ChrNames <- sort(ChrData[,1])

# files for processing one chromosome
temp_1D_binfile_currChr <- paste0(OutDir, '/temp_1D_ALLBins_CurrChr.bed')

for (i in (1:length(ChrNames))) {
	currChr <- ChrNames[i]
	cat(sprintf("\n Processing chromosome for coverage computation: %s ", currChr))

	# get the bins for the current chromosome
	system(paste0("awk \'{if ($1==\"", currChr, "\") {print $0}}\' ", temp_1D_binfile, " > ", temp_1D_binfile_currChr))

	nline <- as.integer(system(paste("cat", temp_1D_binfile_currChr, "| wc -l"), intern = TRUE))
	if (nline == 0) {
		next
	}
	
	# now extract the HiChIP 1D coverage from the input locus pair and contact count file
	# and for the current chromosome
	temp_1D_Coveragefile <- paste0(OutDir, '/temp_1D_coverage.bed')
	system(paste0("awk -F\'[\t]\' -v OFS=\"\t\" \'{if ((NR>1) && ($1==\"", currChr, "\")) {a[$1\"\t\"$2\"\t\"$3]+=$7; a[$4\"\t\"$5\"\t\"$6]+=$7}} END {for (x in a) {print x,a[x]}}\' ", opt$InpFile, " | sort -k1,1 -k2,2n > ", temp_1D_Coveragefile))

	nline <- as.integer(system(paste("cat", temp_1D_Coveragefile, "| wc -l"), intern = TRUE))
	if (nline == 0) {
		next
	}

	# read the bins for the current chromosome
	temp_1D_bindata <- data.table::fread(temp_1D_binfile_currChr, header=F, sep="\t", stringsAsFactors=F)	
	
	# read the HiChIP coverage values for the current chromosome
	temp_1D_CoverageData <- data.table::fread(temp_1D_Coveragefile, header=F, sep="\t", stringsAsFactors=F)

	# append the coverage values of temp_1D_CoverageData 
	# into the bins in temp_1D_bindata
	coverageVec <- rep(0, nrow(temp_1D_bindata))
	ov <- Overlap1D(temp_1D_bindata, temp_1D_CoverageData, uniqov=FALSE)
	coverageVec[ov$A_AND_B] <- temp_1D_CoverageData[ov$B_AND_A, 4]

	# assign the peak information in the 1D bins
	peakvec <- rep(0, nrow(temp_1D_bindata))

	if (!is.null(opt$PeakFile)) {	
		# now read the peak file - supporting both gzipped and normal files
		temp_peakfile <- paste0(OutDir, '/temp_peaks.bed')
		if (tools::file_ext(opt$PeakFile) == "gz") {
			system(paste0("zcat ", opt$PeakFile, " | awk \'{if (($1==\"", currChr, "\") && ($2 ~ /^[0-9]+$/) && ($3 ~ /^[0-9]+$/)) {print $1\"\t\"$2\"\t\"$3}}\' - > ", temp_peakfile))
		} else {
			system(paste0("cat ", opt$PeakFile, " | awk \'{if (($1==\"", currChr, "\") && ($2 ~ /^[0-9]+$/) && ($3 ~ /^[0-9]+$/)) {print $1\"\t\"$2\"\t\"$3}}\' - > ", temp_peakfile))
		}
		temp_PeakData <- data.table::fread(temp_peakfile, header=F, sep="\t", stringsAsFactors=F)
		if (nrow(temp_PeakData) > 0) {
			ov1 <- Overlap1D(temp_1D_bindata, temp_PeakData, boundary=0, offset=0, uniqov=FALSE)
			peakvec[ov1$A_AND_B] <- 1		
		}
	}	# end peak file non-null condition

	# now construct the final data frame
	FinalDF <- cbind.data.frame(temp_1D_bindata, coverageVec, peakvec)
	colnames(FinalDF) <- c('Chr', 'Start', 'End', 'Coverage', 'IsPeak')

	# increment the valid chromosome counter
	valid_chr_count <- valid_chr_count + 1
	if (valid_chr_count == 1) {
		write.table(FinalDF, opt$OutFile, row.names=F, col.names=T, sep="\t", quote=F, append=F)
	} else {
		write.table(FinalDF, opt$OutFile, row.names=F, col.names=F, sep="\t", quote=F, append=T)
	}

}	# end chromosome loop

# remove temporary files
system(paste("rm", temp_1D_binfile))
system(paste("rm", temp_1D_binfile_currChr))
system(paste("rm", temp_1D_Coveragefile))
if (!is.null(opt$PeakFile)) {
	system(paste("rm", temp_peakfile))
}

